Source/WebCore/dom/FragmentDirectiveRangeFinder.cpp

```diff
@@ -26,51 +26,368 @@
 #include "config.h"
 #include "FragmentDirectiveRangeFinder.h"
 
+#include "HTMLAudioElement.h"
+#include "HTMLIFrameElement.h"
+#include "HTMLImageElement.h"
+#include "HTMLMeterElement.h"
+#include "HTMLObjectElement.h"
+#include "HTMLProgressElement.h"
+#include "HTMLStyleElement.h"
+#include "HTMLScriptElement.h"
+#include "HTMLVideoElement.h"
 #include "Chrome.h"
 #include "ChromeClient.h"
+#include "ComposedTreeIterator.h"
 #include "Document.h"
+#include "DocumentType.h"
 #include "Editor.h"
-#include "FindOptions.h"
+#include "Position.h"
 #include "SimpleRange.h"
+#include "TextBoundaries.h"
 #include "TextIterator.h"
 
+#include <dbg_editing.h>
+#include <dbg_dom.h>
+
 namespace WebCore {
 namespace FragmentDirectiveRangeFinder {
 
-static SimpleRange collapseIfRootsDiffer(SimpleRange&& range)
+static bool isSearchInvisible(Node& node)
 {
-    return &range.start.container->rootNode() == &range.end.container->rootNode()
-        ? WTFMove(range) : SimpleRange { range.start, range.start };
+    // TODO: The computed value of its display property is none.
+    
+    // TODO: If the node serializes as void.
+    
+    if (is<HTMLIFrameElement>(node) || is<HTMLImageElement>(node) || is<HTMLMeterElement>(node) || is<HTMLObjectElement>(node) || is<HTMLProgressElement>(node) || is<HTMLStyleElement>(node) || is<HTMLScriptElement>(node) || is<HTMLVideoElement>(node) || is<HTMLAudioElement>(node))
+        return true;
+    
+    // TODO: Is a select element whose multiple content attribute is absent.
+    
+    return false;
 }
 
-static std::optional<SimpleRange> rangeOfString(const String& target, Document& document)
+static bool isNonSearchableSubtree(Node& node)
 {
-    if (target.isEmpty())
+    Node* traversingNode = &node;
+    while(traversingNode) {
+        if (isSearchInvisible(*traversingNode))
+            return true;
+        traversingNode = traversingNode->parentInComposedTree();
+    }
+    return false;
+}
+
+static Node& nearestBlockAncestor(Node& node)
+{
+    Node* curNode = &node;
+    while (curNode)
+    {
+        if (!curNode->isTextNode() && curNode->renderer() && curNode->renderer()->style().isDisplayBlockLevel())
+            return *curNode;
+        else
+            curNode = curNode->parentInComposedTree();
+    }
+    return node.document();
+}
+
+static std::optional<BoundaryPoint> getBoundaryPointAtIndex(unsigned index, Vector<Node*> textNodeList, bool isEnd)
+{
+    unsigned counted = 0;
```

Comment:
Can we find a better name than `counted`? I don't know this area of code very well but maybe `startIndex`? `counted` really doesn't tell me much about what it's counting and what for.